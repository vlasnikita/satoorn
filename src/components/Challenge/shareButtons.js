import React, { Component } from 'react';

export default class ShareButton extends Component {

    constructor(props) {
        super(props)
        this.shareTemplate = props.shareTemplate
    }

   

    render() {
        return (
            <a
                id={this.props.id}
                className="Challenge__shareIcon"
                target="_blank"
                href={this.shareTemplate.apply(this.props.challengeId, this.props.challengeTitle)}>
                <img src={this.shareTemplate.imgSrc} 
                alt={this.shareTemplate.alt} 
                />
            </a>
        )
    }

}