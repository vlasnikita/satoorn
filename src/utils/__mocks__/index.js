const utils = jest.requireActual('Utils/index')

export const getOperationHash = jest.fn().mockImplementation(str => 'MOCKED_HASH')

export const getDonatersEmoji = jest.fn().mockImplementation(() => utils.getDonatersEmoji())

export const linkifyPlainText = jest.fn().mockImplementation((text) => utils.linkifyPlainText(text))

export const openPaymentWidget =  jest.fn().mockImplementation((amount, transactionId, desc, isDonation, challengeId, challengeBody) =>{})