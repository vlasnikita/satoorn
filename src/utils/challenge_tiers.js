import venzelGreenIcon from "Static/challenge/venzel-green.svg";
import venzelBlueIcon from "Static/challenge/venzel-blue.svg";
import venzelPurpleIcon from "Static/challenge/venzel-purple.svg";
import venzelYellowIcon from "Static/challenge/venzel-yellow.svg";
import React from "react";

const TIER_LIMITS_RUR = {
    t5: 0,
    t4: 150,
    t3: 300,
    t2: 700,
    t1: 1000
}

export const getTierNumber = donationAmountString => {
    const donationAmount = +donationAmountString

    if(donationAmount < TIER_LIMITS_RUR.t4) return 5
    else if(donationAmount < TIER_LIMITS_RUR.t3) return 4
    else if(donationAmount < TIER_LIMITS_RUR.t2) return 3
    else if(donationAmount < TIER_LIMITS_RUR.t1) return 2
    else return 1
}

export const getVenzel = donationAmountString => {
    const donationAmount = +donationAmountString
    const tier = getTierNumber(donationAmount);

    if (tier === 4) return <img className='Challenge__venzel' src={venzelGreenIcon} alt=""/>;
    else if (tier === 3) return <img className='Challenge__venzel' src={venzelBlueIcon} alt=""/>;
    else if (tier === 2) return <img className='Challenge__venzel' src={venzelPurpleIcon} alt=""/>;
    else if (tier === 1) return <img className='Challenge__venzel' src={venzelYellowIcon} alt=""/>
};