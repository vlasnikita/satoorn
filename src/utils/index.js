import {UnitPay} from 'Static/js/unitpay'
import {YooMoneyCheckout} from 'Static/js/yoomoney'
import sha256 from 'crypto-js/sha256'
import QRCode  from 'qrcode';

import store from 'Store/index';
import { SECRET_KEY, PUBLIC_KEY, HOST } from 'Constants/payment_module'
import { HOSTING_ADDRESS } from 'Constants/routes'
import ALERT_TYPE from 'Constants/alert_type'

import { setActiveTransactionSuccess, setActiveTransactionFail } from 'Actions/transactions'
import { setRedirectToChallenge } from 'Actions/challenges'
import { updateChallenge } from 'Actions/challenges'
import { showAlert } from 'Actions/alert'
import { getChallengeById } from 'Actions/challenges';
import log from './logging';

export const validateEmail = email => {
    const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(String(email).toLowerCase());
}

export const validateUrl = url => {
    const re = /(((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:)*@)?(((\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5]))|((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.?)(:\d*)?)(\/((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)+(\/(([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)*)*)?)?(\?((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|[\uE000-\uF8FF]|\/|\?)*)?(\#((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|\/|\?)*)?$/
    return re.test(String(url).toLowerCase());
}

export const concatenateUrlParams = paramsMap => {
    let urlParamsString = ''
    const isObject = paramsMap === Object(paramsMap) && Object.prototype.toString.call(paramsMap) !== '[object Array]'
    if(!paramsMap || !isObject) return urlParamsString

    const keys = Object.keys(paramsMap)
    if(keys.length === 0) return urlParamsString
    else {
        Object.keys(paramsMap).forEach(paramKey => {
            if(!!paramsMap[paramKey]) {
                if (urlParamsString.length === 0) urlParamsString = `?${paramKey}=${paramsMap[paramKey]}`
                else urlParamsString = `${urlParamsString}&${paramKey}=${paramsMap[paramKey]}`
            }
        })
        return urlParamsString
    }
}

export const getFormSignature = (amount, transaction, desc) => {
    const hashStr = `${transaction}{up}${desc}{up}${amount}{up}${SECRET_KEY}`
    return sha256(hashStr)
}

export const openInNewTab = url => {
    var win = window.open(url, '_blank');
    win.focus();
}

export const openPaymentWidget = (amount, transaction, desc, isDonation, challengeId, challengeBody, streamerLogin) => {
    const signature = getFormSignature(amount, transaction, desc)

    // store.dispatch(getChallengeById(challengeId))

    const checkout = new window.YooMoneyCheckoutWidget({
        confirmation_token: 'confirmation-token', //TODO set proper token
        return_url: 'https://satoorn.ru/', //TODO set redirect 200 url
        customization: {
            //TODO customize view here
            modal: true
        },
        error_callback: function(error) {
            console.log(error)
        }
    });
  
    checkout.on('modal_close', () => {
        console.log('Код, который нужно выполнить после закрытия всплывающего окна.')
    });

    checkout.on('success', () => {
        log.debug('Успешный платеж', params);

        if(isDonation) {
            store.dispatch(updateChallenge(challengeId, {...challengeBody, isJustUpdated: true}))
        }
        else {
            store.dispatch(setActiveTransactionSuccess())
            store.dispatch(setRedirectToChallenge(challengeId, streamerLogin))
        }
        store.dispatch(showAlert(ALERT_TYPE.OK))

        checkout.destroy();
    });

    checkout.on('fail', () => {
        log.debug(message, params);

        store.dispatch(setActiveTransactionFail())
        store.dispatch(showAlert(ALERT_TYPE.PS_ERROR))

        checkout.destroy();
    });

    //Отображение платежной формы в контейнере
    checkout.render('payment-form');


    /*
    ----- OBSOLETE UNITPAY, EXAMPLE PURPOSES -----
    */
    // const payment = new UnitPay();
    // payment.createWidget({
    //     sum: +amount,
    //     desc: desc.toString(),
    //     account: transaction.toString(),
    //     domainName: "unitpay.money",
    //     signature: signature.toString(),
    //     publicKey: PUBLIC_KEY,
    //     locale: 'ru'
    // });

    // payment.success(function (params) {
    //     log.debug('Успешный платеж', params);

    //     if(isDonation) {
    //         store.dispatch(updateChallenge(challengeId, {...challengeBody, isJustUpdated: true}))
    //     }
    //     else {
    //         store.dispatch(setActiveTransactionSuccess())
    //         store.dispatch(setRedirectToChallenge(challengeId, streamerLogin))
    //     }
    //     store.dispatch(showAlert(ALERT_TYPE.OK))
    // });
    // payment.error(function (message, params) {
    //     log.debug(message, params);

    //     store.dispatch(setActiveTransactionFail())
    //     store.dispatch(showAlert(ALERT_TYPE.PS_ERROR))
    // });
    // return false;
}

export const getDonatersEmoji = () => {
    if(process.env.NODE_ENV === 'test') return ''
    const emojis = ['😱', '😵', '🤩', '🤗','🤔', '😮', '😌', '🙃']
    return emojis[ Math.floor(Math.random() * emojis.length) ]
}

export const getAlertSuccessEmoji = () => {
    if(process.env.NODE_ENV === 'test') return ''
    const emojis = [
        '😊','😋','😉','️😏','😌','🤤', '😸','️🙌',
        '👍','👊','✊','✌', '️🤘','🤟','👌','🤙',
        '🐒','🙈', '🌝','🎉','🎊','🔥','✨']

    return emojis[ Math.floor(Math.random() * emojis.length) ]

}

export const getOperationHash = str => {
    return sha256(str).toString()
};

export const linkifyPlainText = text => text.replace(/((http:|https:)[^\s]+[\w])/g, '<a href="$1" target="_blank">$1</a>')

export const checkIfNoOuterUIPage = () => {
    return (
        window.location.pathname === '/policy' ||
        window.location.pathname === '/agreement'
    )
}

export function onlyUnique(value, index, self) {
    return self.indexOf(value) === index;
}

export const getItemByKey = (array, key) => array.reduce((acc, item) => {
    if(!acc) return item.key === key ? item : null
    else return acc
}, null)

export const generateQRCode = async text => {
    try {
      return await QRCode.toDataURL(text)
    } catch (err) {
      console.error(err)
    }
}