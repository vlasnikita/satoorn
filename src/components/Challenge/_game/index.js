import React, {Component, Fragment} from 'react';
import {connect} from "react-redux";

import {Challenge} from 'Components/Challenge';
import {updateChallenge} from "Actions/challenges";
import {getVenzel} from "Utils/challenge_tiers";
import {closePopup, openPopup} from "Actions/popup";
import {showAlert} from "Actions/alert";
import {setWidgetChallenge} from "Actions/widgetChallenge";

// const API = 'https://api.rawg.io/api/games?page_size=5&page=1&key=163b014b0d7b4fe4b49c79f77ec56157&ordering=-rating&search='

export class ChallengeGame extends Challenge {
    constructor(props) {
        super(props)
    }

    render() {
        const { donationAmount, title } = this.props.challenge
        return (
            <div className="Challenge_game">
                <span className="Challenge_game__number">{this.props.number}</span>
                <div className="Challenge__price">
                    {getVenzel(donationAmount)}
                    <p className={`Challenge__priceText Challenge__priceText_t${this.props.number}`}>
                        {('' + donationAmount).split(/\D/).join('')}
                        <span className="Challenge__priceText_currency">â‚½</span>
                    </p>
                </div>
                <p className="Challenge_game__title">{title}</p>
                {this.getActions()}
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    profile: state.profile.profile,
    configs: state.configs,
    activeStreamer: state.activeStreamer.activeStreamer
});

const mapDispatchToProps = (dispatch) => {
    return {
        handleOpenPopup: (type, acceptingChallenge) => dispatch(openPopup(type, acceptingChallenge)),
        handleClosePopup: () => dispatch(closePopup()),
        handleShowAlert: type => dispatch(showAlert(type)),

        handleSetWidgetChallenge: challenge => dispatch(setWidgetChallenge(challenge)),
        handleUpdateChallenge: (id, body) => dispatch(updateChallenge(id, body))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(ChallengeGame)
