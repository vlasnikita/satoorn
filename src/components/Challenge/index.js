import React, {Component, Fragment} from 'react';
import {connect} from 'react-redux';
import {Link,Redirect} from "react-router-dom";

import {showAlert} from "Actions/alert";
import {updateChallenge} from 'Actions/challenges';
import {setWidgetChallenge} from 'Actions/widgetChallenge';
import {closePopup, openPopup} from 'Actions/popup';
import {setActivePayment} from 'Actions/payment';
import ALERT_TYPE from "Constants/alert_type";
import CHALLENGE_STATE from 'Constants/challenge_states';
import POPUP_TYPE from 'Constants/popup_type';
import {donateChallenge, rejectChallenge} from 'Fetchers';
import {getOperationHash, openPaymentWidget} from 'Utils';
import {numbersMask} from 'Utils/input_masks';
import {logDonateEvent, logShareEvent, logUserAction} from 'Utils/metrics';
import {getTierNumber,getVenzel} from 'Utils/challenge_tiers';
import {FB_TEMPLATE, TG_TEMPLATE, TWITTER_TEMPLATE, VK_TEMPLATE} from 'Constants/share_links';
import ShareButton from './shareButtons';
import log from 'Utils/logging';

import heartIcon from 'Static/challenge/heart.svg'
import obsIcon from 'Static/challenge/obs.png'

import bgWhite from 'Static/challenge/bg-white.png'
import bgGreen from 'Static/challenge/bg-green.png'
import bgBlue from 'Static/challenge/bg-blue.png'
import bgPurple from 'Static/challenge/bg-purple.png'
import bgYellow from 'Static/challenge/bg-yellow.png'

export class Challenge extends Component {

    eventContext = 'CHALLENGE_FEED';
    challengeProps = () => this.props.challenge;
    hashReject = null;
    hashDonate = null;

    constructor(props) {
        super(props);
        this.descriptionNode = React.createRef();
        this.state = {
            newBidAmount: props.configs.minPaymentAmount.value,
            minBid: props.configs.minPaymentAmount.value,
            messageDraft: '',
            isError: false,
            isAgreed: false,
            isBalloonError: false,
            isCanceled: false,
            isInputFocused: false,
            isRedirectToWidgets: false,
            isPaymentMethodsPopupVisible: false
        }
    }

    componentDidMount() {
        this.updateRejectHash();
        this.updateDonateHash();

        const descNode = this.descriptionNode.current;
        if (descNode && descNode.scrollHeight < descNode.offsetHeight + 6) {
            descNode.classList.add('Challenge__description_noellipsis')
        }
        this.removeUpdatedFlagOnComletedUpdate();
    }

    componentDidUpdate() {
        this.removeUpdatedFlagOnComletedUpdate();
        if (this.minAmountConfigDefined()) this.setState({
            newBidAmount: this.props.configs.minPaymentAmount.value,
            minBid: this.props.configs.minPaymentAmount.value,
        })
    }

    removeUpdatedFlag = () => {
        setTimeout(
            () => this.props.handleUpdateChallenge(this.challengeProps().id, {isJustUpdated: false}),
            1250
        )
    };

    expand = e => {
        e.target.style.display = 'none';
        this.descriptionNode.current.style.maxHeight = '9999px'
    };

    getPrice = () => {
        const tier = getTierNumber(this.props.challenge.donationAmount)
        return (
            <div className={`Challenge__cover Challenge__cover_t${tier}`}>
                <span className="Challenge__coverShadow"/>
                {this.getCoverBg(tier)}
                <div className="Challenge__price">
                    {getVenzel(this.props.challenge.donationAmount)}
                    <p className="Challenge__priceText">
                        {('' + this.challengeProps().donationAmount).split(/\D/).join('')}
                        <span className="Challenge__priceText_currency">₽</span>
                    </p>
                </div>
            </div>
        )
    };

    getCoverBg = tier => {
        if(tier == 5) return <img src={bgWhite} />
        else if(tier == 4) return <img src={bgGreen} />
        else if(tier == 3) return <img src={bgBlue} />
        else if(tier == 2) return <img src={bgPurple} />
        else if(tier == 1) return <img src={bgYellow} />

    }

    getActions = () => {
        if (this.isMyChallenge()) {
            if (this.challengeStateIs(CHALLENGE_STATE.ACCEPTED)) {
                return this.getShareButtons()
            }
            else {
                return (
                    <Fragment>
                        {this.getStreamerActions()}
                    </Fragment>
                )
            }
        }
        else {
            if (this.challengeStateIs(CHALLENGE_STATE.ACCEPTED)) return null;
            else {
                return this.renderDefaultChallenge();
            }
        }
    };

    handleDonationChange = e => {
        const newBidAmount = numbersMask(e.target.value);
        this.setState({
            newBidAmount,
            isError: false
        })
    };

    initDonation = () => {
        if (!this.props.profile.id) this.props.handleOpenPopup(POPUP_TYPE.UNREGISTRED);
        else if (this.state.newBidAmount < this.state.minBid) this.setState({isError: true});
        else if (!this.props.configs.settings.value.challengeDonateDetailsConfirmed) this.props.handleOpenPopup(POPUP_TYPE.DONATE_DISCLAIMER);
        else {
            logDonateEvent(this.eventContext, this.props.profile.id);
            const {id, title, donationAmount, streamerLogin} = this.challengeProps();
            this.props.handleSetActivePayment({
                challenge: {id, title, donationAmount, streamerLogin},
                newBidAmount: this.state.newBidAmount,
                hash: this.hashDonate
            })
            this.props.handleOpenPopup(POPUP_TYPE.PAYMENT_METHODS);
        }
    };

    handleCancelChallenge = () => {
        logUserAction(this.eventContext, 'REJECT', this.props.profile.id);
        rejectChallenge(this.challengeProps().id, this.hashReject)
            .then(() => {
                this.updateRejectHash();
                this.setState({isCanceled: true})
            })
            .catch(() => this.challengeProps().handleShowAlert(ALERT_TYPE.INTERNAL_ERROR))

    };

    handleAcceptChallenge = () => {
        logUserAction(this.eventContext, 'ACCEPT', this.props.profile.id);
        if (!this.props.configs.settings.value.challengeAcceptDetailsConfirmed) {
            this.props.handleOpenPopup(POPUP_TYPE.ACCEPT_DISCLAIMER, {id: this.challengeProps().id})
        }
        else {
            this.props.handleOpenPopup(POPUP_TYPE.ACCEPT_CHALLENGE, {
                id: this.challengeProps().id,
                amount: this.challengeProps().donationAmount
            })
        }
    };

    handleOpenWidgets = () => {
        this.setState({ isRedirectToWidgets: true })
        logUserAction(this.eventContext, 'OPEN_WIDGETS', this.props.profile.id);
        this.props.handleSetWidgetChallenge(this.props.challenge)
    }

    updateDonateHash() {
        this.hashDonate = getOperationHash(`ChallengeDonate${+new Date()}${this.challengeProps().id}`);
    }

    updateRejectHash() {
        this.hashReject = getOperationHash(`Challenge${+new Date()}${this.challengeProps().id}`);
    }

    renderDefaultChallenge() {
        return (
            <div className={`Challenge__actions ${this.state.isError && 'Challenge__actions_error'}`}>
                <div className="Challenge__inputContainer">
                    <input
                        value={this.state.newBidAmount}
                        placeholder={this.state.minBid}
                        className="Challenge__input"
                        onChange={this.handleDonationChange}
                        onFocus={() => this.setState({isInputFocused: true})}
                        onBlur={() => this.setState({isInputFocused: false})}
                    />
                    <span className="Challenge__inputCurrency">₽</span>
                </div>
                <button
                    id="donateButton"
                    className="Button Button_purple Challenge__button"
                    onClick={this.initDonation}
                >Повысить
                </button>
                {this.state.isError &&
                <p className="Challenge__actions_errorMessage">Минимальная ставка {this.state.minBid}₽</p>}
            </div>
        )
    }

    removeUpdatedFlagOnComletedUpdate() {
        if (this.challengeProps().isJustUpdated) {
            this.removeUpdatedFlag();
        }
    }

    minAmountConfigDefined() {
        return !this.state.minBid && this.props.configs.minPaymentAmount.loaded
            &&
            !!this.props.configs.minPaymentAmount.value;
    }

    challengeStateIs(state) {
        return this.challengeProps().state === state;
    }

    isMyChallenge() {
        return !!this.props.profile.id
            &&
            this.props.profile.login === this.challengeProps().streamerLogin;
    }

    getShareButtons() {
        return (
            <div className="Challenge__actions Challenge__actions_accepted">
                <div className="Challenge__acceptedBlock">
                    <p className="Challenge__acceptedBlockTitle">Вызов принят!</p>
                    <p className="Challenge__acceptedBlockDescription">Расскажи своим подписчикам</p>
                </div>
                <ul className="Challenge__share">
                    <ShareButton
                        onClick={() => this.handleShareButtonClick('TG')}
                        id="shareTg"
                        challengeId={this.challengeProps().id}
                        shareTemplate={TG_TEMPLATE}
                        challengeTitle={this.challengeProps().title}/>
                    <ShareButton
                        onClick={() => this.handleShareButtonClick('VK')}
                        id="shareVk"
                        challengeId={this.challengeProps().id}
                        shareTemplate={VK_TEMPLATE}
                        challengeTitle={this.challengeProps().title}/>
                    <ShareButton
                        onClick={() => this.handleShareButtonClick('TWITTER')}
                        id="shareTwitter"
                        challengeId={this.challengeProps().id}
                        shareTemplate={TWITTER_TEMPLATE}
                        challengeTitle={this.challengeProps().title}/>
                    <ShareButton
                        onClick={() => this.handleShareButtonClick('FB')}
                        id="shareFb"
                        challengeId={this.challengeProps().id}
                        shareTemplate={FB_TEMPLATE}
                        challengeTitle={this.challengeProps().title}/>
                </ul>
            </div>
        );
    }

    getStreamerActions() {
        return (<div className='Challenge__actions Challenge__actions_borderless' style={{display: this.state.isAgreed ? 'none' : 'flex'}}>
            <button
                className="Button Button_greytext Challenge__button_cancel"
                onClick={this.handleCancelChallenge}
            >Отказаться
            </button>
            <button
                className={`Button Button_purple Challenge__button_accept ${this.props.isWidgets ? 'Challenge__button_setwidget' : ''}`}
                onClick={this.handleAcceptChallenge}
            >{this.props.isWidgets ? '+' : 'Принять'}
            </button>
        </div>)
    }

    handleShareButtonClick(event) {
        logShareEvent(this.eventContext, event, this.props.profile.id);
    }

    performFiatDonate() {
        logDonateEvent(this.eventContext, this.props.profile.id);
        const {id, title, donationAmount} = this.challengeProps();
        donateChallenge(this.state.newBidAmount, id, this.hashDonate)
            .then(res => {
                if (res.code !== ALERT_TYPE.OK) {
                    this.props.handleShowAlert(res.code);
                }
                else {
                    this.updateDonateHash();

                    const desc = `Повышение челленджа ${title}`;

                    log.debug(desc, res);

                    openPaymentWidget(this.state.newBidAmount, res.transactionId, desc, true, id, {donationAmount: +amount + this.state.newBidAmount});
                }
            })
            .catch(() => this.props.handleShowAlert(ALERT_TYPE.INTERNAL_ERROR));
    }

    getMeta = () => {
        return (
            <div className="Challenge__meta">
                <div className="Challenge__metaColumn Challenge__metaColumn_mgRight">
                    <div className="Challenge__metaRow">
                        <div className="Challenge__metaDonaters">
                            <div className="Challenge__metaText">{this.challengeProps().donatedPersonCount || 1}</div>
                            <img src={heartIcon} className="Challenge__metaIcon"/>
                        </div>
                    </div>
                </div>
                <div className="Challenge__metaColumn">
                    <span className="Challenge__metaLabel Challenge__metaLabel_solo">Автор</span>
                    <div className="Challenge__metaRow">
                        <Link
                            to={`/${this.challengeProps().authorLogin}`}
                            className="Challenge__metaText"
                        >{this.challengeProps().authorName}</Link>
                    </div>
                </div>
            </div>
        )
    }

    renderCoverGotos(){
        return(
            <Fragment>
                {this.isMyChallenge() && <button
                    className="Challenge__goto"
                    onClick={this.handleOpenWidgets}
                >
                    <img src={obsIcon} className="Challenge__gotoIcon"/>
                    <span className="Challenge__gotoText">Вставить в OBS/Xsplit</span>
                </button>}
            </Fragment>
        )
    }

    renderPaymentMethods(){

        return (
            <div className="Popup">
                <div className="Popup__body">
                    <button onClick={() => this.setState({ isPaymentMethodsPopupVisible: false })} className="Popup__close">&#x2715;</button>
                    BBBBBBBBB
                </div>
            </div>
        )
    }

    render() {
        if(this.state.isRedirectToWidgets) return <Redirect to={`/${this.props.challenge.streamerLogin}/widgets`}/>
        else if (this.state.isCanceled) return null;
        // noinspection HtmlRequiredAltAttribute
        else return (
            <div
                className={
                    `Challenge 
                    ${this.props.isJustUpdated ? 'Challenge_updated' : ''}
                    ${this.props.isWidgets ? 'Challenge_widgetsFeed' : ''}`
                }
                id={`Challenge_${this.challengeProps().id}`}
            >
                <span
                    className="Challenge__widgetSetter"
                    onClick={() => this.props.isWidgets ? this.props.handleSetWidgetChallenge(this.props.challenge) : null}
                />
                <div className="Challenge__header">
                    <div className="Challenge__streamer">
                        <img src={this.props.challenge.streamerPhotoLink}
                             className="Challenge__streamerAvatar"/>
                        <Link
                            to={`/${this.props.challenge.streamerLogin}`}
                            className="Challenge__streamerLogin">
                            {this.props.challenge.streamerName}
                        </Link>
                    </div>
                    {this.getMeta()}
                </div>
                {this.getPrice()}
                <div className="Challenge__body">
                    <Link
                        className="Challenge__title"
                        to={`/${this.challengeProps().streamerLogin}/challenges/${this.challengeProps().id}`}
                    >{this.challengeProps().title}</Link>
                    {this.getActions()}
                </div>
                <div ref={this.descriptionNode} className="Challenge__description">
                    {this.challengeProps().description}
                    <span className="Challenge__descriptionExpand"/>
                </div>
                <Link
                    className="Challenge__description Challenge__description_link"
                    to={`/challenges/${this.challengeProps().id}`}
                >Открыть челлендж</Link>
                {this.renderCoverGotos()}
                {this.state.isPaymentMethodsPopupVisible && this.renderPaymentMethods()}
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    profile: state.profile.profile,
    configs: state.configs
});

const mapDispatchToProps = (dispatch) => {
    return {
        handleOpenPopup: (type, acceptingChallenge) => dispatch(openPopup(type, acceptingChallenge)),
        handleClosePopup: () => dispatch(closePopup()),
        handleShowAlert: type => dispatch(showAlert(type)),

        handleSetWidgetChallenge: challenge => dispatch(setWidgetChallenge(challenge)),
        handleUpdateChallenge: (id, body) => dispatch(updateChallenge(id, body)),
        handleSetActivePayment: payload => dispatch(setActivePayment(payload))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(Challenge);