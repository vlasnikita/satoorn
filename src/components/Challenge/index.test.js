import {configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import 'jest-specific-snapshot';
import React from 'react';
import {Provider} from "react-redux";
import {BrowserRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';
import configureMockStore from "redux-mock-store";
import * as challengeTestUtils from '../../test/utils/challengeTestUtils';
import Challenge from './index';

configure({adapter: new Adapter()});

const mockStore = configureMockStore();
const initialState = {
    profile: {
        profile: {
            id: '5f108c57399c8300013dad42',
            authenticationSource: 'twitch',
            externalId: '204865822',
            externalProfileLink: null,
            externalPhotoLink: 'https://static-cdn.jtvnw.net/user-default-pictures-uv/75305d54-c7cc-40d1-bb9c-91fbe85943c7-profile_image-300x300.png',
            email: null,
            login: 'vlasnikita',
            bio: null,
            walletAmount: 0,
            viewerCount: 0,
            followerCount: 0,
            challengeCount: 0,
            streamer: false,
            contacts: []
        },
        loaded: true
    },
    configs:{
        minPaymentAmount: {
            value: 30
        }
    }
};

it('renders correctly someone\'s Challenge', () => {

    function createNodeMock(element) {
        if (element.className === 'Challenge__description') {
            // This is your fake DOM node for <p>.
            // Feel free to add any stub methods, e.g. focus() or any
            // other methods necessary to prevent crashes in your components.
            return {};
        }
        // You can return any object from this method for any type of DOM component.
        // React will use it as a ref instead of a DOM node when snapshot testing.
        return element
    }

    const store = mockStore(initialState);
    const options = {createNodeMock};
    const challenge = require('Mocks/challenges')[0];
    const tree = renderer.create(
        <BrowserRouter>
            <Provider store={store}>
                <Challenge
                    id={challenge.id}
                    challenge={challenge}
                />
            </Provider>
        </BrowserRouter>,
        options
    );

    expect(tree.toJSON()).toMatchSnapshot();
});


it('should renders correctly accepted challenge', () => {
    const challenge = challengeTestUtils.renderDefaultChallengeFeedItem(
        undefined,
        challengeTestUtils.myAcceptedChallengeData
    );
    expect(toJson(challenge)).toMatchSpecificSnapshot('./__specifics__/acceptedChallengeFeedItem.specific')
});

it('should renders correctly my challenge', () => {
    const challenge = challengeTestUtils.renderDefaultChallengeFeedItem(
        undefined,
        challengeTestUtils.myChallengeData
    );
    expect(toJson(challenge)).toMatchSpecificSnapshot('./__specifics__/myChallengeFeedItem.specific')
});