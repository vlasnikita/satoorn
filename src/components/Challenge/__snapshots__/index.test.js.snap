// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly someone's Challenge 1`] = `
<div
  className="Challenge 
                    
                    "
  id="Challenge_1"
>
  <span
    className="Challenge__widgetSetter"
    onClick={[Function]}
  />
  <div
    className="Challenge__header"
  >
    <div
      className="Challenge__streamer"
    >
      <img
        className="Challenge__streamerAvatar"
        src="https://images-cdn.9gag.com/photo/anbZVN0_700b.jpg"
      />
      <a
        className="Challenge__streamerLogin"
        href="/user_1"
        onClick={[Function]}
      />
    </div>
    <div
      className="Challenge__meta"
    >
      <div
        className="Challenge__metaColumn Challenge__metaColumn_mgRight"
      >
        <div
          className="Challenge__metaRow"
        >
          <div
            className="Challenge__metaDonaters"
          >
            <div
              className="Challenge__metaText"
            >
              10
            </div>
            <img
              className="Challenge__metaIcon"
              src="heart.svg"
            />
          </div>
        </div>
      </div>
      <div
        className="Challenge__metaColumn"
      >
        <span
          className="Challenge__metaLabel Challenge__metaLabel_solo"
        >
          Автор
        </span>
        <div
          className="Challenge__metaRow"
        >
          <a
            className="Challenge__metaText"
            href="/akareinaka"
            onClick={[Function]}
          >
            akareinaka
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="Challenge__cover Challenge__cover_t5"
  >
    <span
      className="Challenge__coverShadow"
    />
    <img
      src="bg-white.png"
    />
    <div
      className="Challenge__price"
    >
      <p
        className="Challenge__priceText"
      >
        78
        <span
          className="Challenge__priceText_currency"
        >
          ₽
        </span>
      </p>
    </div>
  </div>
  <div
    className="Challenge__body"
  >
    <a
      className="Challenge__title"
      href="/user_1/challenges/1"
      onClick={[Function]}
    >
      some title
    </a>
    <div
      className="Challenge__actions false"
    >
      <div
        className="Challenge__inputContainer"
      >
        <input
          className="Challenge__input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder={30}
          value={30}
        />
        <span
          className="Challenge__inputCurrency"
        >
          ₽
        </span>
      </div>
      <button
        className="Button Button_purple Challenge__button"
        id="donateButton"
        onClick={[Function]}
      >
        Повысить
      </button>
    </div>
  </div>
  <div
    className="Challenge__description"
  >
    Some default description Some default description Some default description Some default description Some default description Some default description Some default description Some default description Some default description Some default description 
    <span
      className="Challenge__descriptionExpand"
    />
  </div>
  <a
    className="Challenge__description Challenge__description_link"
    href="/challenges/1"
    onClick={[Function]}
  >
    Открыть челлендж
  </a>
</div>
`;
